sdk/python/feast/pyspark/historical_feature_retrieval_job.py

```diff
@@ -1,19 +1,433 @@
+        """
+        feast_to_spark_type_mapping = {
+            "bytes": "binary",
+            "string": "string",
+            "int32": "int",
+            "int64": "bigint",
+            "double": "double",
+            "float": "float",
+            "bool": "boolean",
+            "bytes_list": "array<binary>",
+            "string_list": "array<string>",
+            "int32_list": "array<int>",
+            "int64_list": "array<bigint>",
+            "double_list": "array<double>",
+            "float_list": "array<float>",
+            "bool_list": "array<boolean>",
+        }
+        return feast_to_spark_type_mapping[self.type.lower()]
+
+
+class FeatureTable(NamedTuple):
+    """
+    Feature table specification.
+
+    Attributes:
+        name (str): Table name.
+        entities (List[Field]): Primary keys for the features.
+        features (List[Field]): Feature list.
+        max_age (int): In seconds. determines the lower bound of the timestamp of the retrieved feature.
+            If not specified, this would be unbounded
+        project (str): Feast project name.
+    """
+
+    name: str
+    entities: List[Field]
+    features: List[Field]
+    max_age: Optional[int] = None
+    project: Optional[str] = None
+
+    @property
+    def entity_names(self):
+        """
+        Returns columns names for the entities.
+        """
+        return [field.name for field in self.entities]
+
+    @property
+    def feature_names(self):
+        """
+        Returns columns names for the features.
+        """
+        return [field.name for field in self.features]
+
+
+class FileDestination(NamedTuple):
+    """
+    Output file for the spark job.
+
+    Attributes:
+        format (str): Output format.
+        path (str): Output uri.
+    """
+
+    format: str
+    path: str
+
+
+class EntityDataframe(NamedTuple):
+    """
+    Entity dataframe with specification.
+
+    Attributes:
+        df (DataFrame): Dataframe for the table.
+        timestamp_column (str): Column representing the event timestamp.
+    """
+
+    df: DataFrame
+    timestamp_column: str
+
+    @classmethod
+    def from_source(cls, spark: SparkSession, source: Source):
+        """
+        Construct an EntityDataframe instance based on the Source.
+
+        Args:
+            spark (SparkSession): Spark session.
+            source (Source): Entity table source.
+
+        Returns:
+            EntityDataframe: Instance of EntityDataframe.
+        """
+
+        df = (
+            spark.read.format(source.spark_format)
+            .options(**source.spark_read_options)
+            .load(source.spark_path)
+        )
+
+        mapped_entity_df = _map_column(df, source.mapping)
+
+        return EntityDataframe(
+            df=mapped_entity_df, timestamp_column=source.timestamp_column
+        )
+
+
+class FeatureTableDataframe(NamedTuple):
+    """
+    Feature table dataframe with specification.
```

Comment:
Can you add a comment here that describes this class. This comment says nearly nothing.

Reply:
I dont understand why we need this class

Reply:
This class contains information from both FeatureTable and Source, excluding the location of the feature table. In my original implementation, both timestamp columns and feature table specification are within the same level of configuration. But for the implementation in ingestion job, timestamp columns is in Source, whereas feature table specification is in FeatureTable.

I can see two choices here:
1. Eliminate FeatureTableDataframe, and pass both Source and FeatureTable into `as_of_join` and `join_entity_to_feature_tables`, eventhough the two methods don't exactly need to know the format and path of the Feature Table source.

2. Move timestamp columns to FeatureTable instead of Source, but that would mean the configuration is different from the Ingestion Job.

Reply:
> This class contains information from both FeatureTable and Source, excluding the location of the feature table. In my original implementation, both timestamp columns and feature table specification are within the same level of configuration. But for the implementation in ingestion job, timestamp columns is in Source, whereas feature table specification is in FeatureTable.

Ok, but it's not clear what this class is for. You're talking about your past implementations and implementation details but what problem are you trying to solve. This class is called FeatureTableDataframe, could it be replaced by using a FeatureTable and a DataFrame?


> I can see two choices here:
> 
> 1. Eliminate FeatureTableDataframe, and pass both Source and FeatureTable into `as_of_join` and `join_entity_to_feature_tables`, eventhough the two methods don't exactly need to know the format and path of the Feature Table source.
> 2. Move timestamp columns to FeatureTable instead of Source, but that would mean the configuration is different from the Ingestion Job.

You are presenting solutions but I dont know what problem you are trying to solve.

Reply:
Feature table class doesn't contain timestamp column and created timestamp column. If it does, then yes, I don't need this class and I can easily replace this with a FeatureTable and a Dataframe, which is my preferred solution. The only reason why I haven't done this is because of the previous discussion where we want to use similar configuration for ingestion job.

The problem I am trying to solve here: I want  neither as_of_join method nor join_entity_features method to have any argument that contains file format and file path information, so that means I can't pass Source as an argument. But timestamp column information is required. Without creating a new class, that means my function arguments for as of join methods, will need to have list of Feature Table, List or dictionary of event timestamp columns, and the dataframes. The new class acts as a container that store all these methods together.

That being say, I don't like this class myself either. I will move the timestamp column to FeatureTable and remove FeatureTableDataframe.